src/wp-includes/canonical.php

```diff
@@ -976,8 +976,9 @@ function redirect_guess_404_permalink() {
 		}
 
 		$publicly_viewable_statuses = array_filter( get_post_stati(), 'is_post_status_viewable' );
+		$publicly_viewable_post_types = array_filter( get_post_types(), 'is_post_type_viewable' );
 		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
-		$post_id = $wpdb->get_var( "SELECT ID FROM $wpdb->posts WHERE $where AND post_status IN ('" . implode( "', '", esc_sql( $publicly_viewable_statuses ) ) . "')" );
+		$post_id = $wpdb->get_var( $wpdb->prepare( "SELECT ID FROM $wpdb->posts WHERE $where AND post_status IN (%s) AND post_type IN (%s)", implode( "', '", esc_sql( $publicly_viewable_statuses ) ), implode( "', '", esc_sql( $publicly_viewable_post_types ) ) ) );
```

Comment:
The `%s` replacement doesn't support `IN` queries, this change will result in escaped quotes, eg `post_type IN ('\'post\', \'page\'')` so will never return any results. It will look for a single post type with the slug `'post', 'page'` which won't exist.

I saw your [screen shot that my original patch caused an error](https://prnt.sc/UM-8i_df528e), so I'll see where I goofed -- it's possibly because I duplicated the `post_type IN` clause.